go/test/endtoend/sharding/initialsharding/sharding_util.go

```diff
@@ -0,0 +1,676 @@
+}
+
+func initClusterForInitialSharding(ksName string, shardNames []string, totalTabletsRequired int, rdonly bool, isMulti bool) {
+	var keyspace cluster.Keyspace
+	var ksExists bool
+	if ClusterInstance.Keyspaces == nil {
+		keyspace = cluster.Keyspace{
+			Name: ksName,
+		}
+	} else {
+		// this is single keyspace test, so picking up 0th index
+		ksExists = true
+		keyspace = ClusterInstance.Keyspaces[0]
+	}
+
+	for _, shardName := range shardNames {
+		shard := &cluster.Shard{
+			Name: shardName,
+		}
+
+		for i := 0; i < totalTabletsRequired; i++ {
+			// instantiate vttablet object with reserved ports
+			tabletUID := ClusterInstance.GetAndReserveTabletUID()
+			tablet := &cluster.Vttablet{
+				TabletUID: tabletUID,
+				HTTPPort:  ClusterInstance.GetAndReservePort(),
+				GrpcPort:  ClusterInstance.GetAndReservePort(),
+				MySQLPort: ClusterInstance.GetAndReservePort(),
+				Alias:     fmt.Sprintf("%s-%010d", ClusterInstance.Cell, tabletUID),
+			}
+			if i == 0 { // Make the first one as master
+				tablet.Type = "master"
+			} else if i == totalTabletsRequired-1 && rdonly { // Make the last one as rdonly if rdonly flag is passed
+				tablet.Type = "rdonly"
+			}
+			// Start Mysqlctl process
+			tablet.MysqlctlProcess = *cluster.MysqlCtlProcessInstance(tablet.TabletUID, tablet.MySQLPort, ClusterInstance.TmpDirectory)
+			if !isMulti {
+				if err := tablet.MysqlctlProcess.StartWithArgs(path.Join(os.Getenv("VTROOT"), "config", "mycnf", "rbr.cnf")); err != nil {
+					return
+				}
+			} else {
+				tablet.MysqlctlProcess.InitDBFile = newInitDbFile
+				if err := tablet.MysqlctlProcess.StartWithArgs(path.Join(os.Getenv("VTROOT"), "config", "mycnf", "rbr.cnf"), "-db-credentials-file", dbCredentialFile); err != nil {
+					return
+				}
+			}
+
+			// start vttablet process
+			tablet.VttabletProcess = cluster.VttabletProcessInstance(tablet.HTTPPort,
+				tablet.GrpcPort,
+				tablet.TabletUID,
+				ClusterInstance.Cell,
+				shardName,
+				ksName,
+				ClusterInstance.VtctldProcess.Port,
+				tablet.Type,
+				ClusterInstance.TopoProcess.Port,
+				ClusterInstance.Hostname,
+				ClusterInstance.TmpDirectory,
+				ClusterInstance.VtTabletExtraArgs,
+				ClusterInstance.EnableSemiSync)
+			tablet.Alias = tablet.VttabletProcess.TabletPath
+			tablet.VttabletProcess.DbPwd = dbPwd
+			tablet.VttabletProcess.EnableSemiSync = true
+			shard.Vttablets = append(shard.Vttablets, *tablet)
+		}
+		if ksExists {
+			ClusterInstance.Keyspaces[0].Shards = append(ClusterInstance.Keyspaces[0].Shards, *shard)
+		} else {
+			keyspace.Shards = append(keyspace.Shards, *shard)
+		}
+	}
+	if len(ClusterInstance.Keyspaces) == 0 {
+		ClusterInstance.Keyspaces = append(ClusterInstance.Keyspaces, keyspace)
+	}
+}
+
+// TestInitialShardingWithVersion - main test with accepts different params for various test
+func TestInitialShardingWithVersion(t *testing.T, keyspace *cluster.Keyspace, shardingKeyType topodata.KeyspaceIdType, isMulti bool, isExternal bool) {
```

Comment:
we are no longer doing v2, so rename this to just `TestInitialSharding`?
Instead of using KeyspaceIdType (uint64 or bytes), maybe we should be using the type of the column directly here. KeyspaceIdType is a v2 concept.

Reply:
To accommodate `bytes[]` and reuse existing test, I passed this as parameter.

Reply:
I understand. My point is:
v2 tests use ShardingColumnType, so this had to be set to UINT64 or BYTES. v3 tests don't need to use that specific type. The distinction in v3 is the column type in the tablet itself - whether it is int or varbinary. We can use some other flag to denote that, it should not be tied to ShardingColumnType. Either it can be a string ("integer" vs "string"), or it can be https://github.com/vitessio/vitess/blob/master/go/sqltypes/type.go#L134

Reply:
Ok, I will do that. I was using that as flag, I have not used it to set any sharding column.